src/installer.ts

```diff
@@ -77,20 +80,57 @@ async function acquireGo(version: string): Promise<string> {
     throw new Error('Temp directory not set');
   }
 
-  if (osPlat == 'win32') {
+  // gotip is always .tar.gz.
+  if (osPlat == 'win32' && !gotip) {
     extPath = await tc.extractZip(downloadPath);
   } else {
     extPath = await tc.extractTar(downloadPath);
   }
 
+  let toolRoot = extPath;
+  if (!gotip) {
+    toolRoot = path.join(toolRoot, 'go');
+    version = normalizeVersion(version);
+  }
+
+  //
+  // Build gotip from source.
+  //
+  if (gotip) {
+    // Version embed in the executable. Required for build.
+    let versionFile = path.join(toolRoot, 'VERSION');
+    fs.writeFileSync(versionFile, `gotip-${version}`);
+
+    const options = {
+      cwd: path.join(toolRoot, 'src'),
+      env: {
+        ...process.env,
+        // Embed in the executable, set it explicitly for reproducible builds.
+        'GOROOT_FINAL': 'go',
```

Comment:
I guess I should be more clear on my question; what is this for? I don't need to set this to have a working build of Go. This is the first time I've seen this variable before.

Reply:
Default `GOROOT` value baked into Go binaries is either `GOROOT_FINAL` (if set) or build directory. I’d expect the latter to differ between builds on CI system. Go toolchain uses the `GOROOT` env var that we export explicitly, not the default one.

I’m keen on [reproducible builds](https://reproducible-builds.org) and Go team does an excellent job supporting these. Though we’d also have to compile with `CGO_ENABLED=0` for complete reproducibility, because otherwise external linker and Cgo embeds timestamps / paths in binaries. Note that disabling Cgo during compilation does not entirely disable Cgo, but instead just builds Go without using Cgo.

Again, there is no point in caching build outputs if they depend on the environment. And we want to cache gotip, as the build time is pretty slow compared to an average Go package (1m30s vs few seconds).

You could try running `GOROOT_FINAL=go CGO_ENABLED=0 ./make.bash` in different directories and the results of `shasum ../bin/go` should be the same.

I also recommend using `-trimpath` flag with Go ≥1.13 for building user packages (not the toolchain), i.e. set `GOFLAGS=-trimpath`.

```
-trimpath
	remove all file system paths from the resulting executable.
	Instead of absolute file system paths, the recorded file names
	will begin with either "go" (for the standard library),
	or a module path@version (when using modules),
	or a plain import path (when using GOPATH).
```

Reply:
What should be done is allowing overrides from `process.env`.

```js
env: {
  'GOROOT_FINAL': 'go',
  'CGO_ENABLED': '0',
  ...process.env,
}
```

Reply:
Personally, I wouldn't want any of these options set, as they would not normally be set in a typical Go compiler build. Opting into this behavior if you really want the Go compiler to have these things set is fine, but I'd really want to be as close to a typical build (or the downloaded builds as with the other versions in this action) as possible. I'm personally not too convinced that having the actual Go compiler be reproducible is a requirement for having the things it builds be reproducible; that seems like the actual use case of reproducible builds in CI.

Can you even set environment variables on pre-created actions? I thought you could only use `with`.

Reply:
>Can you even set environment variables on pre-created actions (without adding extra steps)? I thought you could only use with.

Haven’t tried yet, but why wouldn’t that work?

>Commands run in actions or steps can create, read, and modify environment variables.

https://help.github.com/en/actions/automating-your-workflow-with-github-actions/using-environment-variables

Reply:
Not really what I meant.

```yaml
- use: actions/setup-go@master
  with:
    go-version: tip
  env:
    CGO_ENABLED: '0'
```

etc

What would environment variables mean for an action like this which is supposed to be caching results? Especially if the toolcache library is just taking in `version` as its key?

I feel like if a user has a specific requirement for the Go toolchain other than its version, they're going to have to do something special themselves with the knowledge that a generalized action may not have all of the functionality they require by default (i.e. "if you want to change what the compiler is built with, no caching for you", making things work for the general user who has no knowledge of setup-go's internals).

Reply:
My point is that if you are using Cgo, you probably already have `CGO_ENABLED=1` env exported to clearly state that this package uses Cgo. It’s really, really confusing when large projects uses  Cgo dependencies and does not explicitly declare itself as using Cgo, because you can’t just `grep` for `import "C"`.

And GitHub Actions doc (a bit vaguely) states that `env` should work for pre-created actions.

Reply:
And adding `GOROOT_FINAL=go` just makes the build reproducible. That’s it.

Reply:
GOROOT_FINAL doesn't seem harmful, so I'm indifferent either way, I'm just saying that it's not a strict requirement for having an action that itself can produce a reproducible build.

> if you are using Cgo, you probably already have `CGO_ENABLED=1` env exported to clearly state that this package uses Cgo. 

I have not seen anyone do this, especially since the default is enabled. It's really not important for installing Go in this action, either way.

Reply:
>I have not seen anyone do this

That’s the problem.

I’ve clearly stated my reasoning above. Summing up the main points,

- if you are using Cgo, you probably already have `CGO_ENABLED=1`,
- there is no point in caching build outputs if they depend on the environment.

If you don’t want Cgo to be disabled by default (if that’s your objection), we can export `CGO_ENABLED=$($GOROOT_BOOTSTRAP/bin/go env CGO_ENABLED)` after the build. But I strongly oppose the implicit Cgo.

Also, there are no packages using `tip` version with [actions/setup-go@master](https://github.com/actions/setup-go). So we won’t break anything. And if you do indeed depend on some quirky Cgo feature, be it the external linker behavior, or plain `import "C"`, and want to test against `tip`, you are probably prepared for things to explode.

Reply:
I really can't agree. All disabling cgo is going to do is piss off every user of the action who (knowingly or not) depend on it, i.e. anyone with many database drivers like sqlite, embedded luajit, etc, and cause unnecessary "why doesn't this work" churn. I get that the average "tip" user in CI is going to know better, but I don't feel that needs to be related, especially if people want to switch from any other CI platform to Actions and discover "oh, this action is opinionated differently than every other CI so now I've got to do this other thing".

"I have not seen anyone do this" being a problem you're trying to solve, and "you probably already have GCO_ENABLED=1" are two competing thoughts. They cannot both be true.

> Also, there are no packages using tip version with actions/setup-go@master.

Well yeah, the functionality doesn't exist... Other than my previous comments which build it (and I run tip on my projects so I can report issues upstream).

Either way, your PR does not include any cgo changes, so whatever.

Reply:
To be clear, I'm just talking about the environment _after_ setup-go. If you want to build the Go compiler without cgo, then that's really not a major problem (other than it differing from running `make.bash` from a typical environment, and not necessarily being needed for said compiler to produce a consistent output; maybe it's necessary for proper caching between builds). I just don't want to see things getting quietly set for future steps in the pipeline.

Reply:
Sadly, looks like we won’t reach an agreement on this, as we both seem to have a strong opinion on the topic.

>I have not seen anyone do this

Sorry for being unclear in the previous comment, I meant that the problem is with packages not defining `CGO_ENABLED=1` in the environment. Personally, I have a negative experience working working with such projects, so my opinion is biased.

Also, I’ve seen people doing matrix builds for both `CGO_ENABLED=1` and `0`.

That said, maybe we can cache iff `CGO_ENABLED=0` and provide an opt-in for caching (non-reproducible) Cgo builds? Looks like good compromise to me.

Reply:
I don't understand what "packages not defining `CGO_ENABLED=1`" means; Go packages cannot specify the build environment they're intended to be built within. If it breaks, it breaks. Packages which intend to be built with either or need to guard themselves with build tags.

> Also, I’ve seen people doing matrix builds for both `CGO_ENABLED=1` and `0`.

This to me is evidence that setup-go should not set these variables for future steps (and only use internally as needed), as matrix builds set their environments at the start. An action overriding them defeats that.

Again, if you want to set `CGO_ENABLED=0` strictly for `make.bash`/`make.bat` as a part of building Go at `tip` and caching it between workers, then that's a different story, and I can understand wanting that to be consistent.

Reply:
Ah, maybe GitHub will add automatic updates for comments one day… Didn’t see the follow ups. Anyway, if disabling caches for `CGO_ENABLED=1` is not the problem, we can agree on this. Otherwise, it’s possible to export `CGO_ENABLED` from bootstrap toolchain for future steps and have a faster CI runs with Cgo.

Reply:
>Again, if you want to set CGO_ENABLED=0 strictly for make.bash/make.bat as a part of building Go at tip and caching it between workers, then that's a different story, and I can understand wanting that to be consistent.

Then we can finally settle on `CGO_ENABLED=$($GOROOT_BOOTSTRAP/bin/go env CGO_ENABLED)` for future steps and `CGO_ENABLED=0` for `make.bash`/`make.bat`. Yay!

Reply:
```
CGO_ENABLED=$($GOROOT_BOOTSTRAP/bin/go env CGO_ENABLED)
```

I'm assuming you're talking about `::set-env` here; isn't this a no-op?

Reply:
It’s is, but only if using such matrix builds.

If `CGO_ENABLED` is not defined, then we default to whatever bootstrap toolchain sees fit for the host platform. Otherwise, as you’ve said earlier, we’d might change the default behavior of the toolchain.